### PR TITLE
Optimized types

### PR DESCRIPTION
**Pull request by AntoineRestivo**
Please check the definition of the input types before merge

### CODE REVIEW

File: src/BosonSampling.jl [hunk 1]
```diff
@@ -37,13 +40,10 @@ include("scattering.jl")
 include("bunching/bunching.jl")
 include("partitions/legacy.jl")
 
-include("circuits/circuit_elements.jl")
```
Comment (benoitseron) [inline src/BosonSampling.jl:40]: why did you remove this? it is necessary I believe for my work on bunching

File: src/distributions/noisy_distribution.jl [hunk 2]
```diff
@@ -7,8 +7,9 @@ function noisy_distribution(;input, distinguishability, reflectivity, interf, ex
     δ = 1e-4
 
     input_modes = input.r.state
-    number_photons = input.r.n
-    number_modes = input.r.m
+    number_photons = input.n
+    number_modes = input.m
+    distinguishability = input.distinguishability
```
Comment (benoitseron) [inline src/distributions/noisy_distribution.jl:12]: see comments on distinguishability in inputs

File: src/distributions/theoretical_distribution.jl [hunk 3]
```diff
@@ -1,10 +1,12 @@
-function theoretical_distribution(;input::Input, distinguishability::Real, interf::Interferometer, gram_matrix::GramMatrix, i=nothing)
+function theoretical_distribution(;input::Input, interf::Interferometer, i=nothing)
 
     input_modes = input.r.state
-    number_photons = input.r.n
-    number_modes = input.r.m
+    number_photons = input.n
+    number_modes = input.m
+    distinguishability = input.distinguishability
```
Comment (benoitseron) [inline src/distributions/theoretical_distribution.jl:6]: same as before

File: src/example_usage.jl [hunk 4]
```diff
@@ -8,7 +8,7 @@ m = 5
 n = 3
 
 interf = RandHaar(m)
-i = Input{PartDist}(first_modes(n,m), GramMatrix{PartDist}(n, rand_gram_matrix(n)))
+i = Input{RandomModel}(first_modes(n,m))
```
Comment (benoitseron) [inline src/example_usage.jl:11]: See comments on input types

File: src/types/input.jl [hunk 5]
```diff
@@ -2,81 +2,151 @@
...
-struct PartDist <:InputType
-end
-abstract type ToyModel <: InputType
-end
-abstract type RandomModel <: InputType
-end
-struct Distinguishable <:InputType
+abstract type Bosonic <: InputType end
+
+abstract type GaussianState <: Bosonic end
```
Comment (benoitseron) [inline src/types/input.jl:7]: bosonic is meant to say perfectly indistinguishable so this subtyping shouldn't be

File: src/types/input.jl [hunk 6]
```diff
@@ -2,81 +2,151 @@
...
-struct Bosonic <:InputType
-end
-struct PartDist <:InputType
-end
-abstract type ToyModel <: InputType
-end
-abstract type RandomModel <: InputType
-end
-struct Distinguishable <:InputType
+abstract type Bosonic <: InputType end
```
Comment (benoitseron) [inline src/types/input.jl:5]: shouldn't be an abstract type, see comment below

File: src/types/input.jl [hunk 7]
```diff
@@ -2,81 +2,151 @@
...
+struct ToyModel <: PartDist
+    distinguishability::Float64
+    ToyModel(distinguishability::Float64) = new(distinguishability)
 end
 
+abstract type RandomModel <: PartDist end
+
+abstract type Undef <: InputType end
+
+abstract type Distinguishable <: InputType end
```
Comment (benoitseron) [inline src/types/input.jl:39]: you shouldn't make these concrete stuff abstract types

File: src/types/input.jl [hunk 8]
```diff
@@ -2,81 +2,151 @@
...
             new(nothing)
         else
             is_orthonormal(vectors_matrix, atol = ATOL) ? new(vectors_matrix) : error("invalid orthonormal basis")
         end
     end
+
 end
 
-struct GramMatrix{T<:InputType}
+mutable struct GramMatrix{T<:InputType}
```
Comment (benoitseron) [inline src/types/input.jl:58]: I'm not sure putting stuff mutable for this is good for performance, the gram matrix shouldn't change anyways and the previous code accounted that

File: src/types/input.jl [hunk 9]
```diff
@@ -2,81 +2,151 @@
...
 end
 
 struct Input{T<:InputType}
+
     r::ModeOccupation
-    G::GramMatrix
+    G::Matrix
     n::Int
     m::Int
+    distinguishability::Any
```
Comment (benoitseron) [inline src/types/input.jl:122]: don't use Any, rather Union{Number, Nothing} for stuff like this

then, distinguishability is not something easy to define, I'll explain when we see each other - there is still debate if there is even a measure of it so I believe this should be removed

File: src/types/input.jl [hunk 10]
```diff
@@ -2,81 +2,151 @@
...
         end
     end
+
 end
 
 struct Input{T<:InputType}
+
     r::ModeOccupation
-    G::GramMatrix
+    G::Matrix
```
Comment (benoitseron) [inline src/types/input.jl:119]: why? if you do this then some code must break if we use the gram matrix structure

**Review by benoitseron — CHANGES_REQUESTED** [review]
So in general, try not to change stuff that might break code like 
- changing properties of stuff (concrete to abstract types, GramMatrix to Matrix...)
- We need to discuss the specific naming of input types etc to be as close as possible with the literature/not lead into mistakes new users

Otherwise the rest seems good

File: src/types/input.jl (hunk 9 above)
Comment (AntoineRestivo) [reply src/types/input.jl:122]: I agree, the point is that all algorithms we currently have (sampler + computation of the statistics) are based on this interpolating model. And the latter can still be easily generalised wether you do not want to use a continuous parameter for distinguishability

File: src/types/input.jl (hunk 6 above)
Comment (AntoineRestivo) [reply src/types/input.jl:5]: If this is not the case then we cannot state that ToyModel or RandomModel are subtypes of ParDist and in this way I do not understand the point of defining ParDist if we define them as subtypes of InputType

File: src/types/input.jl (hunk 10 above)
Comment (benoitseron) [reply src/types/input.jl:119]: see discussion

File: src/distributions/noisy_distribution.jl (hunk 2 above)
Comment (benoitseron) [reply src/distributions/noisy_distribution.jl:12]: one_parameter_distinguishability
